controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -1038,11 +1038,16 @@ func (r *BareMetalHostReconciler) getHardwareDetailsFromAnnotation(host *metal3v
 		return nil, nil
 	}
 	content := []byte(annotations[hardwareDetailsAnnotation])
-	objHardwareDetails := &metal3v1alpha1.HardwareDetails{}
-	if err := json.Unmarshal(content, objHardwareDetails); err != nil {
+	objHardwareDetails := metal3v1alpha1.HardwareDetails{}
+	if err := json.Unmarshal(content, &objHardwareDetails); err != nil {
```

Comment:
I think the code could be simplified by using the `DisallowUnknownFields()` of the json Decoder, ie just: 

```
decoder := json.NewDecoder(strings.NewReader(annotations[hardwareDetailsAnnotation]))
decoder.DisallowUnknownFields()

if err := decoder.Decode(&objHardwareDetails); err != nil {
	return nil, err
}
```

In this way if the annotation string contains a non-matching key in the target struct an error should be returned

Reply:
Thanks, I switched to this approach in the latest version